File: .agents/skills/codebase-intelligence/scripts/typescript/hotspots.js
#!/usr/bin/env node
// hotspots.js — detects high-risk files: churn (git) × complexity
// Usage: node hotspots.js --repo <abs-path> --out <abs-path-to-json>

'use strict';

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const REFACTOR_THRESHOLD = 50;

// --- arg parsing ---
const argv = process.argv.slice(2);
let repoPath = null;
let outPath = null;

for (let i = 0; i < argv.length; i++) {
  if (argv[i] === '--repo') repoPath = argv[++i];
  else if (argv[i] === '--out') outPath = argv[++i];
}

if (!repoPath || !outPath) {
  const msg = { layer: 'hotspots', lang: 'typescript', error: true, error_message: '--repo and --out are required' };
  process.stdout.write(JSON.stringify(msg, null, 2) + '\n');
  process.exit(0);
}

fs.mkdirSync(path.dirname(outPath), { recursive: true });

const outDir = path.dirname(outPath);
const rawDir = path.join(outDir, 'raw', 'typescript');

// --- Step 1: git log churn ---
let gitLogOutput = '';
try {
  gitLogOutput = execSync(
    'git log --since="90 days ago" --name-only --pretty=format: HEAD',
    { cwd: repoPath, encoding: 'utf8', stdio: ['pipe', 'pipe', 'pipe'] }
  );
} catch (err) {
  const errMsg = (err.stderr || err.message || '').toLowerCase();
  if (errMsg.includes('not a git repository') || errMsg.includes('not found')) {
    const out = { layer: 'hotspots', lang: 'typescript', error: true, error_message: `git log failed: ${err.message}` };
    fs.writeFileSync(outPath, JSON.stringify(out, null, 2));
    process.exit(0);
  }
  // Use partial output if available
  gitLogOutput = err.stdout || '';
}

// Filter and count TS/JS files
const tsJsPattern = /\.(ts|tsx|js|jsx)$/;
const churnMap = {};
let totalLines = 0;

for (const rawLine of gitLogOutput.split('\n')) {
  const line = rawLine.trim();
  if (!line) continue;
  totalLines++;
  if (tsJsPattern.test(line)) {
    churnMap[line] = (churnMap[line] || 0) + 1;
  }
}

// Emit a warning if fewer than 30 total file entries (sparse history)
const warnings = [];
if (totalLines < 30) {
  warnings.push(`Sparse git history: only ${totalLines} file-change entries in last 90 days. Hotspot scores may not be reliable.`);
}

// --- Step 2: read complexity index ---
const rawComplexityPath = path.join(rawDir, 'complexity.json');
let complexityByFile = {};
try {
  if (fs.existsSync(rawComplexityPath)) {
    complexityByFile = JSON.parse(fs.readFileSync(rawComplexityPath, 'utf8'));
  }
} catch (e) {
  // non-fatal — use default max_cyclomatic = 1
}

// --- Step 3: compute scores ---
const hotspots = [];

for (const [filePath, churnCount] of Object.entries(churnMap)) {
  const maxCyclomatic = complexityByFile[filePath] || 1;
  const score = churnCount * maxCyclomatic;
  hotspots.push({
    path: filePath,
    commits_last_90d: churnCount,
    max_cyclomatic: maxCyclomatic,
    score,
    verdict: score >= REFACTOR_THRESHOLD ? 'refactor_target' : 'watch',
  });
}

// Sort by score descending
hotspots.sort((a, b) => b.score - a.score);

const refactorTargets = hotspots.filter(h => h.verdict === 'refactor_target');
const topFile = hotspots.length > 0 ? hotspots[0] : null;

const output = {
  layer: 'hotspots',
  lang: 'typescript',
  findings: hotspots,
  counts: {
    refactor_targets: refactorTargets.length,
  },
  summary: {
    top_file: topFile ? topFile.path : null,
    top_score: topFile ? topFile.score : null,
    warnings,
  },
};

fs.writeFileSync(outPath, JSON.stringify(output, null, 2));
process.exit(0);

File: .agents/skills/codebase-intelligence/scripts/python/hotspots.py
#!/usr/bin/env python3
"""Layer 6 — Hotspots (churn × complexity).

A "hotspot" is a file that is BOTH changing often AND complex. These are
the highest-leverage refactor targets: investing here pays compound interest
because the team keeps editing them.

Formula:
    score = log(1 + commits_last_90d) * max_cyclomatic_in_file

Files with score above the 90th percentile are flagged as `refactor_target`.

If the repo is a shallow clone (fewer than 30 commits), this layer warns
and emits an empty findings list — the user should run `git fetch --unshallow`.
"""
from __future__ import annotations
import argparse
import json
import math
import subprocess
import sys
from pathlib import Path
from collections import Counter

SHALLOW_THRESHOLD = 30
WINDOW_DAYS = 90


def git_total_commits(repo: Path) -> int:
    try:
        out = subprocess.check_output(
            ["git", "rev-list", "--count", "HEAD"], cwd=str(repo), text=True
        )
        return int(out.strip())
    except (subprocess.CalledProcessError, FileNotFoundError, ValueError):
        return 0


def git_churn(repo: Path) -> Counter:
    """Count commits per .py file in the last WINDOW_DAYS."""
    try:
        out = subprocess.check_output(
            ["git", "log", f"--since={WINDOW_DAYS}.days.ago",
             "--name-only", "--pretty=format:", "--", "*.py"],
            cwd=str(repo), text=True
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return Counter()

    counter: Counter = Counter()
    for line in out.splitlines():
        line = line.strip()
        if line and line.endswith(".py"):
            counter[line] += 1
    return counter


def max_cc_per_file(repo: Path) -> dict[str, int]:
    """Run `radon cc -j` and return max complexity per file."""
    args = ["radon", "cc", str(repo), "-j",
            "-e", "*/migrations/*,*/.venv/*,*/venv/*,*/build/*,*/dist/*"]
    try:
        result = subprocess.run(args, capture_output=True, text=True, check=False)
    except FileNotFoundError:
        return {}
    if not result.stdout.strip():
        return {}
    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError:
        return {}

    out: dict[str, int] = {}
    for file_path, items in data.items():
        if not isinstance(items, list):
            continue
        try:
            rel = str(Path(file_path).resolve().relative_to(repo))
        except ValueError:
            rel = file_path
        complexities = [it.get("complexity", 0) for it in items]
        if complexities:
            out[rel] = max(complexities)
    return out


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("repo")
    ap.add_argument("--targets", required=False)
    ap.add_argument("--out", required=True)
    args = ap.parse_args()

    repo = Path(args.repo).resolve()
    out = Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)

    total_commits = git_total_commits(repo)
    if total_commits == 0:
        payload = {
            "layer": "hotspots",
            "skipped": True,
            "reason": "not a git repository or git not available",
            "findings": [],
            "counts": {"total": 0},
        }
        out.write_text(json.dumps(payload, indent=2))
        print("hotspots: skipped (no git history)")
        return 0
    if total_commits < SHALLOW_THRESHOLD:
        payload = {
            "layer": "hotspots",
            "skipped": True,
            "reason": f"shallow clone ({total_commits} commits) — run `git fetch --unshallow`",
            "findings": [],
            "counts": {"total": 0},
        }
        out.write_text(json.dumps(payload, indent=2))
        print(f"hotspots: skipped (shallow clone, {total_commits} commits)")
        return 0

    churn = git_churn(repo)
    cc = max_cc_per_file(repo)

    scored: list[dict] = []
    for path, commits in churn.items():
        max_cc = cc.get(path, 1)
        score = round(math.log1p(commits) * max_cc, 2)
        scored.append({
            "path": path,
            "commits_last_90d": commits,
            "max_cyclomatic": max_cc,
            "score": score,
        })
    scored.sort(key=lambda x: x["score"], reverse=True)

    # Mark the top 10% (or top 5, whichever is smaller) as refactor targets
    cutoff = max(5, int(len(scored) * 0.1)) if scored else 0
    for i, item in enumerate(scored):
        item["verdict"] = "refactor_target" if i < cutoff else "informational"

    payload = {
        "layer": "hotspots",
        "window_days": WINDOW_DAYS,
        "total_commits_in_repo": total_commits,
        "findings": scored[:50],  # cap at top 50 to keep JSON small
        "counts": {
            "total": len(scored),
            "refactor_targets": min(cutoff, len(scored)),
        },
    }
    out.write_text(json.dumps(payload, indent=2))
    top = scored[0]["path"] if scored else "—"
    print(f"hotspots: {len(scored)} files scored, top risk = {top}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
